cms/djangoapps/contentstore/proctoring.py

```diff
@@ -0,0 +1,127 @@
+"""
+Code related to the handling of Proctored Exams in Studio
+"""
+
+import logging
+
+from django.conf import settings
+
+from xmodule.modulestore.django import modulestore
+
+from edx_proctoring.api import (
+    get_exam_by_content_id,
+    update_exam,
+    create_exam,
+    get_all_exams_for_course,
+)
+from edx_proctoring.exceptions import (
+    ProctoredExamNotFoundException
+)
+
+log = logging.getLogger(__name__)
+
+
+def _is_in_course_tree(item):
+    """
+    Check that the item is in the course tree.
+
+    It's possible that the item is not in the course tree
+    if its parent has been deleted and is now an orphan.
+    """
+    ancestor = item.get_parent()
+    while ancestor is not None and ancestor.location.category != "course":
+        ancestor = ancestor.get_parent()
+
+    return ancestor is not None
+
+
+def register_proctored_exams(course_key):
+    """
+    This is typically called on a course published signal. The course is examined for sequences
+    that are marked as timed exams. Then these are registered with the edx-proctoring
+    subsystem. Likewise, if formerly registered exams are unmarked, then those
+    registred exams are marked as inactive
+    """
+
+    if not settings.FEATURES.get('ENABLE_PROCTORED_EXAMS'):
+        # if feature is not enabled then do a quick exit
+        return
+
+    course = modulestore().get_course(course_key)
+    if not course.enable_proctored_exams:
+        # likewise if course does not have this feature turned on
+        return
+
+    # get all sequences, since they can be marked as timed/proctored exams
+    _timed_exams = modulestore().get_items(
+        course_key,
+        qualifiers={
+            'category': 'sequential',
+        },
+        settings={
+            'is_time_limited': True,
+        }
+    )
+
+    # filter out any potential dangling sequences
```

Comment:
@wedaly brought this up. It might be a legacy artifact where pieces of a course could be left 'untetherd' from it's parent, due to a deletion operation. I don't think get_items() does the inspection to see if there is still a parent node that points to the matched item.

Not sure if is still an issue these days, but this extra filtering step can't hurt.